### PR TITLE
`LDAPStorageProvider` count queries

### LINKED ISSUE #20156: Revisit count queries for LDAPStorageProvider
https://github.com/keycloak/keycloak/pull/19050#discussion_r1184810342

### PR DESCRIPTION
**Pull request by vramik**
Closed https://github.com/keycloak/keycloak/issues/20156

### CODE REVIEW

**Comment by mposolda** [comment]
@vramik @mhajas It seems this PR should be rebased. Once that is done and only this commit appears in this PR https://github.com/keycloak/keycloak/commit/759098007ce200b7257751ebc478dbdb2b81714b, you can consider this PR as approved from myself :-)

**Comment by mhajas** [comment]
Thank you @mposolda, unfortunately, our assumption was wrong and it seems this PR breaks the `query` method in `UserStorageManager`. This code:
https://github.com/keycloak/keycloak/blob/34a68d935497af489ceec5e6c541995092147991/model/legacy-private/src/main/java/org/keycloak/storage/UserStorageManager.java#L243-L258
assumes the first provider in the stream can provide enough users to pass `firstResult` number of users so we can skip them, however, this is no longer true after setting a fixed value in the count query.

I came up with another solution idea (note this is not 100% correct I just wanted to pitch an idea on how to fix it) and now waiting for @vramik for his opinion on that:
https://github.com/mhajas/keycloak/commit/2685e0ee541921d26e32592d424468b21ef22064

**Comment by vramik** [comment]
> I came up with another solution idea (note this is not 100% correct I just wanted to pitch an idea on how to fix it) and now waiting for @vramik for his opinion on that:
> [mhajas@2685e0e](https://github.com/mhajas/keycloak/commit/2685e0ee541921d26e32592d424468b21ef22064)

The idea lgtm. We've agreed I'll cherry-pick the commit and update documentation and add a test.

**Review by mhajas — CHANGES_REQUESTED** [review]
Thank you @vramik!

Could you please also mention the new capability interfaces in: https://github.com/keycloak/keycloak/blob/main/model/legacy/src/main/java/org/keycloak/storage/UserStorageProvider.java#L31

Also, I added a few comments into the code.

**Comment by vramik** [comment]
@mhajas thank you for the comments, could you please take another look?

File: server-spi/src/main/java/org/keycloak/storage/user/UserQueryMethods.java [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+     * Searches all users in the realm.
+     *
+     * @param realm a reference to the realm.
+     * @return a non-null {@link Stream} of users.
+     * @deprecated Use {@link #searchForUserStream(RealmModel, Map)} with an empty params map instead.
+     */
+    @Deprecated
+    default Stream<UserModel> getUsersStream(RealmModel realm) {
+        return searchForUserStream(realm, Collections.emptyMap());
+    }
```
Comment (hmlnarik) [inline server-spi/src/main/java/org/keycloak/storage/user/UserQueryMethods.java:51]: Should the deprecated methods be kept in the new interface or rather remove them?

File: server-spi/src/main/java/org/keycloak/storage/user/UserQueryMethods.java [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+ * <code>UserStorageProvider</code> that supports complex user querying. You must
+ * implement this interface if you want to view and manage users from the administration console.
+ * <p/>
+ * Note that all methods in this interface should limit search only to data available within the storage that is
+ * represented by this provider. They should not lookup other storage providers for additional information.
+ *
+ * @author <a href="mailto:bill@burkecentral.com">Bill Burke</a>
+ * @version $Revision: 1 $
+ */
+public interface UserQueryMethods {
```
Comment (hmlnarik) [inline server-spi/src/main/java/org/keycloak/storage/user/UserQueryMethods.java:40]:
```suggestion
public interface UserQueryMethodsProvider {
```

**Review by mhajas — COMMENTED** [review]
Thanks, I added one phrasing comment to the doc.

Also, I would like to hear an opinion from someone else before we merge. @hmlnarik @mposolda Are any of you available for another opinion on that?

**Review by hmlnarik — COMMENTED** [review]
Thanks for this PR. It looks reasonable, though I've added few comments inline to minor issues.

File: server-spi/src/main/java/org/keycloak/storage/user/UserQueryMethods.java (hunk 1 above)
Comment (vramik) [reply server-spi/src/main/java/org/keycloak/storage/user/UserQueryMethods.java]: I'd like to keep it here for the sake of this PR and remove deprecated methods in separate issue (https://github.com/keycloak/keycloak/issues/20438), which also has milestone 22.0.0. There is draft https://github.com/keycloak/keycloak/pull/20440 sent.

**Review by mhajas — APPROVED** [review]
Thanks @vramik, LGTM.

**Review by hmlnarik — APPROVED** [review]
Thank you, @vramik and @mhajas!

**Review by mposolda — APPROVED** [review]
This approach works for me, Thanks!

**Pull request merged**